### PR TITLE
Move serialization logic into Serializer and CollectionSerializer

### PR DESCRIPTION
**Pull request by bf4**
- Remove most Attributes adapter recursion
- Moving Attributes#serializable_hash_for_single_resource to Serializer#serialize

Extracted from #1684

### CODE REVIEW

File: lib/active_model/serializer.rb [hunk 1]
```diff
@@ -98,6 +98,16 @@ def self.get_serializer_for(klass)
       end
     end
 
+    def self.include_directive_from_options(options)
```
Comment (bf4) [inline lib/active_model/serializer.rb]: missing `@api private`.  These methods are being held here for now, but may move around, be renamed, or disappear

**Comment by NullVoxPopuli** [comment]
@bf4 are you wanting to add a changelog entry?

anywho, the change looks good. I'd forgotten we had some serializer stuff in an adapter.

File: lib/active_model/serializer.rb [hunk 2]
```diff
@@ -151,9 +162,9 @@ def success?
     #     serializer.as_json(include: { posts: { include: { comments: { only: :body } }, only: :title } })
     def serializable_hash(adapter_opts = nil)
       adapter_opts ||= {}
-      adapter_opts = { include: '*', adapter: :attributes }.merge!(adapter_opts)
-      adapter = ActiveModelSerializers::Adapter.create(self, adapter_opts)
-      adapter.serializable_hash(adapter_opts)
+      adapter_opts = { include: '*' }.merge!(adapter_opts)
+      adapter_instance = ActiveModelSerializers::Adapter::Attributes.new(self, adapter_opts)
+      serialize(adapter_opts, {}, adapter_instance)
```
Comment (bf4) [inline lib/active_model/serializer.rb]: It really irritates me that this requires an adapter_instance (else I get a stack overflow).  

The adapter_instance interface is

- cached_name : used in making cache key
- class : used by fragment caching in fragment serializers
- fragment_cache : used by fragment caching

File: lib/active_model_serializers/adapter/attributes.rb [hunk 3]
```diff
@@ -1,65 +1,9 @@
...
-        opts = instance_options.merge(include_directive: @include_directive[association.key])
-        relationship_value = Attributes.new(association.serializer, opts).serializable_hash(options)
-
-        if association.options[:polymorphic] && relationship_value
-          polymorphic_type = association.serializer.object.class.name.underscore
-          relationship_value = { type: polymorphic_type, polymorphic_type.to_sym => relationship_value }
-        end
-
-        relationship_value
+        serializer.serialize(instance_options, options, self)
```
Comment (bf4) [inline lib/active_model_serializers/adapter/attributes.rb]: thought I intend the `serialize` method as a placeholder for `serializable_hash`, it's a huge polymorphism win that both Serializer and CollectionSerializer implement it

File: lib/active_model/serializer.rb (hunk 2 above)
Comment (bf4) [inline lib/active_model/serializer.rb]: actually, all of those can be class methods, now that I think about it.

File: lib/active_model/serializer.rb [hunk 4]
```diff
@@ -151,9 +166,8 @@ def success?
     #     serializer.as_json(include: { posts: { include: { comments: { only: :body } }, only: :title } })
     def serializable_hash(adapter_opts = nil)
       adapter_opts ||= {}
-      adapter_opts = { include: '*', adapter: :attributes }.merge!(adapter_opts)
-      adapter = ActiveModelSerializers::Adapter.create(self, adapter_opts)
-      adapter.serializable_hash(adapter_opts)
+      adapter_opts = { include: '*' }.merge!(adapter_opts)
```
Comment (bf4) [inline lib/active_model/serializer.rb]: should use the JSONAPI default include directive, I think

File: lib/active_model/serializer.rb [hunk 5]
```diff
@@ -98,6 +98,25 @@ def self.get_serializer_for(klass)
...
+      end
+    end
+
+    def self.serialization_adapter_instance
+      @serialization_adapter_instance ||= ActiveModelSerializers::Adapter::Attributes
+    end
+
+    def self.serialization_include_directive
+      @serialization_include_directive ||= { include: '*' }
+    end
```
Comment (bf4) [inline lib/active_model/serializer.rb]: considered `{ inclue: JSONAPI::IncludeDirective.new('*', allow_wildcard: true) }` but didn't work. :(

File: lib/active_model/serializer.rb [hunk 6]
```diff
@@ -98,6 +98,22 @@ def self.get_serializer_for(klass)
       end
     end
 
+    # @api private
+    def self.include_directive_from_options(options)
```
Comment (bf4) [inline lib/active_model/serializer.rb]: this should probably go in AMS or some other serialization-related class

**Comment by bf4** [comment]
Merging just to keep momentum going

**Pull request merged**